ldelf/ta_elf_rel.c

```diff
@@ -194,23 +233,47 @@ static void e32_process_dyn_rel(const Elf32_Sym *sym_tab, size_t num_syms,
 				const char *str_tab, size_t str_tab_size,
 				Elf32_Rel *rel, Elf32_Addr *where)
 {
-	size_t sym_idx = 0;
 	const char *name = NULL;
 	vaddr_t val = 0;
-	size_t name_idx = 0;
+
+	e32_get_sym_name(sym_tab, num_syms, str_tab, str_tab_size, rel, &name);
+	resolve_sym(name, &val);
+	*where = val;
+}
+
+static void e32_tls_get_module(const Elf32_Sym *sym_tab, size_t num_syms,
+			       const char *str_tab, size_t str_tab_size,
+			       Elf32_Rel *rel, struct ta_elf **mod)
+{
+	const char *name = NULL;
+	size_t sym_idx = 0;
 
 	sym_idx = ELF32_R_SYM(rel->r_info);
 	if (sym_idx >= num_syms)
 		err(TEE_ERROR_BAD_FORMAT, "Symbol index out of range");
 	sym_idx = confine_array_index(sym_idx, num_syms);
+	if (!sym_idx || sym_tab[sym_idx].st_shndx != SHN_UNDEF) {
+		/* No symbol, or symbol is defined in current module */
+		return;
+	}
 
-	name_idx = sym_tab[sym_idx].st_name;
-	if (name_idx >= str_tab_size)
-		err(TEE_ERROR_BAD_FORMAT, "Name index out of range");
-	name = str_tab + name_idx;
+	*mod = NULL;
+	e32_get_sym_name(sym_tab, num_syms, str_tab, str_tab_size, rel, &name);
+	ta_elf_resolve_sym2(name, NULL, mod, NULL);
+	if (!*mod)
+		err(TEE_ERROR_BAD_FORMAT,
+		    "Symbol %s not found for TLS relocation", name);
+}
 
-	resolve_sym(name, &val);
-	*where = val;
+static void e32_tls_resolve(const Elf32_Sym *sym_tab, size_t num_syms,
+			    const char *str_tab, size_t str_tab_size,
+			    Elf32_Rel *rel, vaddr_t *val)
+{
+	struct ta_elf *mod = NULL;
+	const char *name = NULL;
+
+	e32_get_sym_name(sym_tab, num_syms, str_tab, str_tab_size, rel, &name);
+	ta_elf_resolve_sym2(name, val, &mod, NULL);
```

Comment:
It seems that `&mod` doesn't need to be supplied for `ta_elf_resolve_sym2()` to work if we're not going to use that value.

Reply:
So true! Now fixed.